src/emulator/auth/operations.ts

```diff
@@ -230,6 +231,157 @@ function lookup(
+          existingEmails.add(userInfo.email);
+        }
+      }
+    }
+
+    // Check that there is no duplicate (providerId, rawId) tuple.
+    const existingProviderAccounts = new Set<string>();
+    for (const userInfo of reqBody.users) {
+      for (const { providerId, rawId } of userInfo.providerUserInfo ?? []) {
+        const key = `${providerId}:${rawId}`;
+        assert(
+          !existingProviderAccounts.has(key),
+          `DUPLICATE_RAW_ID : Provider id(${providerId}), Raw id(${rawId})`
+        );
+        existingProviderAccounts.add(key);
+      }
+    }
+  }
+
+  if (!reqBody.allowOverwrite) {
+    const existingLocalIds = new Set<string>();
+    for (const userInfo of reqBody.users) {
+      const localId = userInfo.localId || "";
+      assert(!existingLocalIds.has(localId), `DUPLICATE_LOCAL_ID : ${localId}`);
+      existingLocalIds.add(localId);
+    }
+  }
+
+  const errors: { index: number; message: string }[] = [];
+  for (let index = 0; index < reqBody.users.length; index++) {
+    const userInfo = reqBody.users[index];
+
+    try {
+      assert(userInfo.localId, "localId is missing");
+      const uploadTime = new Date();
+      const fields: Omit<Partial<UserInfo>, "localId"> = {
+        displayName: userInfo.displayName,
+        photoUrl: userInfo.photoUrl,
+        lastLoginAt: userInfo.lastLoginAt,
+      };
+
+      // password
+      if (userInfo.passwordHash) {
+        // TODO: Check and block non-emulator hashes.
+        fields.passwordHash = userInfo.passwordHash;
+        fields.salt = userInfo.salt;
+        fields.passwordUpdatedAt = uploadTime.getTime();
+      } else if (userInfo.rawPassword) {
+        fields.salt = userInfo.salt || "fakeSalt" + randomId(20);
+        fields.passwordHash = hashPassword(userInfo.rawPassword, fields.salt);
+        fields.passwordUpdatedAt = uploadTime.getTime();
+      }
+
+      // custom attrs
+      if (userInfo.customAttributes) {
+        validateSerializedCustomClaims(userInfo.customAttributes);
+        fields.customAttributes = userInfo.customAttributes;
+      }
+
+      // federated
+      if (userInfo.providerUserInfo) {
+        fields.providerUserInfo = [];
+        for (const providerUserInfo of userInfo.providerUserInfo) {
+          const { providerId, rawId } = providerUserInfo;
+          // These providers are handled automatically by create / update.
+          if (providerId !== PROVIDER_PASSWORD && providerId !== PROVIDER_PHONE) {
+            assert(
+              providerId && rawId,
+              "((Auth Emulator only supports providerUserInfo with providerId AND rawId))"
+            );
+            const existingUserWithProviderRawId = state.getUserByProviderRawId(providerId, rawId);
+            assert(
+              !existingUserWithProviderRawId,
+              `((Auth Emulator does not support importing duplicate Provider id(${providerId}), Raw id(${rawId})))`
```

Comment:
This is not the production error message. I'll add a test and fix this before I merge.